app/src/org/commcare/android/nfc/NfcReadActivity.java

```diff
@@ -0,0 +1,119 @@
+ * will attempt to read the tag. If the read is successful, the value read will be stored as an
+ * extra in the activity's result intent under the key "odk_intent_data", where it can be accessed
+ * by other activities. Otherwise, an appropriate error toast will be shown.
+ *
+ * @author Aliza Stone
+ */
+@TargetApi(Build.VERSION_CODES.JELLY_BEAN)
+public class NfcReadActivity extends NfcActivity {
+
+    private String valueRead;
+    private String[] acceptableTypes;
+
+    @Override
+    protected void initFields() {
+        super.initFields();
+        this.acceptableTypes = parseTypes(getIntent());
+    }
+
+    protected static String[] parseTypes(Intent i) {
+        String typesString = i.getStringExtra(NFC_PAYLOAD_MULT_TYPES_ARG);
+        if (typesString == null || typesString.equals("")) {
+            String singleType = i.getStringExtra(NFC_PAYLOAD_SINGLE_TYPE_ARG);
+            if (singleType == null || singleType.equals("")) {
+                return null;
+            } else {
+                return new String[]{ singleType };
+            }
+        } else {
+            return typesString.split(" ");
+        }
+    }
+
+    @Override
+    protected boolean requiredFieldsMissing() {
+        if (this.acceptableTypes == null || this.acceptableTypes.length == 0) {
+            finishWithErrorToast("nfc.read.no.type");
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    protected void dispatchActionOnTag(Tag tag) {
+        readFromNfcTag(tag);
+    }
+
+    private void readFromNfcTag(Tag tag) {
+        Ndef ndefObject = Ndef.get(tag);
+        try {
+            ndefObject.connect();
+            NdefMessage msg = ndefObject.getNdefMessage();
+            if (msg == null) {
+                finishWithErrorToast("nfc.read.no.data");
+                return;
+            }
+            NdefRecord firstRecord = msg.getRecords()[0];
+            Pair<String, Boolean> resultAndSuccess =
+                    NdefRecordUtil.readValueFromRecord(firstRecord, this.acceptableTypes,
+                            this.domainForType);
+            if (resultAndSuccess.second) {
+                this.valueRead = resultAndSuccess.first;
+                finishWithToast("nfc.read.success", true);
+            } else {
+                finishWithErrorToast(resultAndSuccess.first);
+            }
+        } catch (IOException e) {
+            finishWithErrorToast("nfc.read.io.error");
+        } catch (FormatException e) {
+            finishWithErrorToast("nfc.read.msg.malformed");
+        } finally {
+            try {
+                ndefObject.close();
+            } catch (IOException e) {
+                // nothing we can do
+            }
+        }
+    }
+
+    @Override
+    protected void setResultExtrasBundle(Intent i, boolean success) {
+        Bundle responses = new Bundle();
+        responses.putString("nfc_read_result", success ? "success" : "failure");
```

Comment:
Is this used anywhere? Pattern seems slightly awkward

Reply:
Mm I guess for the read it may be a little unnecessary since you would just know by whether or not you received a value. But it definitely serves a real purpose for the write actions, since there's otherwise no way to know in the model whether or not it worked. Do you think it would be better if I removed this for reads and only had it for writes?

Reply:
I was thinking why not use a boolean?

Reply:
(offlined about this)